tango/integrations/beaker/executor.py

```diff
@@ -0,0 +1,575 @@
+                nonlocal successful, failed
+                try:
+                    exc = future.exception()
+                    if exc is None:
+                        successful.add(step_name)
+                    else:
+                        log_exception(exc, logger)
+                        failed.add(step_name)
+                except concurrent.futures.TimeoutError as exc:
+                    log_exception(exc, logger)
+                    failed.add(step_name)
+
+            return future_done_callback
+
+        update_steps_to_run()
+
+        try:
+            with concurrent.futures.ThreadPoolExecutor(
+                max_workers=self.parallelism or None
+            ) as pool:
+                while steps_to_run or step_futures:
+                    # Submit steps left to run.
+                    for step_name in steps_to_run:
+                        future = pool.submit(
+                            self._execute_sub_graph_for_step, step_graph, step_name, run_name, True
+                        )
+                        future.add_done_callback(make_future_done_callback(step_name))
+                        step_futures.append(future)
+                        submitted_steps.add(step_name)
+
+                    # Wait for something to happen.
+                    _, not_done = concurrent.futures.wait(
+                        step_futures, return_when=concurrent.futures.FIRST_COMPLETED
+                    )
+
+                    # Update the list of running futures.
+                    step_futures.clear()
+                    step_futures = list(not_done)
+
+                    # Update the step queue.
+                    update_steps_to_run()
+        except (KeyboardInterrupt, CancellationError):
+            if step_futures:
+                cli_logger.warning("Received interrupt, canceling steps...")
+                self._is_cancelled.set()
+                concurrent.futures.wait(step_futures)
+            raise
+        finally:
+            self._is_cancelled.clear()
+
+        # NOTE: The 'done callback' added to each future is executed in a thread,
+        # and so might not complete before the last 'update_steps_to_run()' is called
+        # in the loop above. Therefore we have to call 'update_steps_to_run()'
+        # one last time here to ensure the 'not_run' set is up-to-date.
+        update_steps_to_run()
+
+        return ExecutorOutput(successful=successful, failed=failed, not_run=not_run)
+
+    def execute_sub_graph_for_step(
+        self, step_graph: StepGraph, step_name: str, run_name: Optional[str] = None
+    ) -> None:
+        self._execute_sub_graph_for_step(step_graph, step_name, run_name)
+
+    def _check_if_cancelled(self):
+        if self._is_cancelled.is_set():
+            raise RunCancelled
+
+    def _execute_sub_graph_for_step(
+        self,
+        step_graph: StepGraph,
+        step_name: str,
+        run_name: Optional[str],
+        in_thread: bool = False,
```

Comment:
In essence, `in_thread` is analogous to `called_by_executor` in the `__main__` script, correct? Maybe at some point, I should refactor `multicore` to handle it cleanly like this too.

Reply:
Yeup